flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapUnionFunction.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericArrayData;
+import org.apache.flink.table.data.MapData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.types.KeyValueDataType;
+
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#MAP_UNION}. */
+@Internal
+public class MapUnionFunction extends BuiltInScalarFunction {
+
+    private final ArrayData.ElementGetter keyElementGetter;
+    private final ArrayData.ElementGetter valueElementGetter;
+
+    public MapUnionFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.MAP_UNION, context);
+        KeyValueDataType outputType =
+                ((KeyValueDataType) context.getCallContext().getOutputDataType().get());
+        keyElementGetter =
+                ArrayData.createElementGetter(outputType.getKeyDataType().getLogicalType());
+        valueElementGetter =
+                ArrayData.createElementGetter(outputType.getValueDataType().getLogicalType());
+    }
+
+    public @Nullable MapData eval(@Nullable MapData map1, @Nullable MapData map2) {
+        if (map1 == null || map2 == null) {
+            return null;
+        }
+        MapDataForMapUnion resultMap = new MapDataForMapUnion(map1, map2);
+        return resultMap;
+    }
+
+    private class MapDataForMapUnion implements MapData {
+        private ArrayData keysArray;
+        private ArrayData valuesArray;
+        private final Map<Object, Object> map;
+
+        public MapDataForMapUnion(MapData map1, MapData map2) {
+
+            this.map = new HashMap<>();
```

Comment:
`HashMap` under the hood uses java's `equals` which is not 100% same as SQL. 
We need to use SQL with `equalityEvaluator` like for instance done for `ARRAY_DISTINCT`

Reply:
but if use equalityEvaluator, the time complexity will become to n square. because we cannot use hashmap. Is that ok?

Reply:
first we should care about correctness and only then about speed

Reply:
And I check the source code. if the data can pass to this function, they are all binaryData, so we can use java's equal.

Reply:
Can you please elaborate where have you seen guarantees that binaryData's equals is in sync with SQL's sync?
Also there is no limitations to use non binarydata base on current method signature, so we can not rely on it

Reply:
To be honest, I don't know API and SQL there are different. Why binaryData's equals is not sync with SQL's sync.
But If we transfer the input Data to the Expression, the data we received in MapUnionFunction are binaryMap.

Reply:
I check the isEqual method in ARRAY_DISTINCT, but I cannot find it's source code. So I don't know the background logical of IsEqual.

Reply:
> Why binaryData's equals is not sync with SQL's sync?

Because the input might not only be binary data but also generic data. And the two cannot be compared. Also SQL has specifics such as leading/trailing zeros and many other specifics that the code generated equals/hashcodes can handle.

Reply:
fixed it.